internal/store/interfaces.go

```diff
@@ -32,39 +22,10 @@ type StatusTrackingGateway interface {
 // Gateway describes a gateway.
 type Gateway interface {
 	ID() core.GatewayID
-	Meta() map[string]string
-	ShouldUpdate(other Gateway) bool
-	Listeners() []Listener
-	ShouldBind(route Route) bool
-}
-
-// ListenerConfig contains the common configuration
-// options of a listener.
-type ListenerConfig struct {
-	Name     string
-	Hostname string
-	Port     int
-	Protocol string
-	TLS      core.TLSParams
-}
-
-// RouteTrackingListener is an optional extension
-// to Listener that tracks when routes have been
-// bound to it.
-type RouteTrackingListener interface {
-	Listener
-
-	OnRouteAdded(route Route)
-	OnRouteRemoved(id string)
-}
-
-// Listener describes the basic methods of a gateway
-// listener.
-type Listener interface {
-	ID() string
-	CanBind(ctx context.Context, route Route) (bool, error)
-	Config() ListenerConfig
-	IsValid() bool
+	Bind(ctx context.Context, route Route) []string
+	Remove(ctx context.Context, id string) error
+	Resolve() core.ResolvedGateway
```

Comment:
So, rather than calling `CanBind` on a listener, storing that result externally somewhere and then invoking `Resolve` on the route with the corresponding listener that could be bound. The new thing is basically calling `Bind` on the gateway with the route, storing that internally in the gateway state, and then just invoking `Resolve` directly on the gateway to grab the fully resolved state of the gateway that we need to sync over.

When the Binder itself gets modified a bit in the next PR, the `Bind` call gets fully extracted to a `Binder` interface to keep the `Gateway` a bit lighter.